diagnose_cluster.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python
+
+import argparse
+import re
+import json
+from subprocess import call, Popen, PIPE
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--name', '-n', required=True, type=str, help="Cluster name")
+parser.add_argument('--dest', '-d', required=True, type=str, help="Directory for diagnose output -- must be local")
+parser.add_argument('--hail-log', '-l', required=False, type=str, default='/home/hail/hail.log', help="Path for hail.log file")
+parser.add_argument('--overwrite', required=False, action='store_true', help="Delete dest directory before adding new files")
+parser.add_argument('--no-diagnose', required=False, action='store_true', help="Do not run gcloud dataproc clusters diagnose")
+parser.add_argument('--compress', '-z', required=False, action='store_true', help="GZIP all files")
+parser.add_argument('--workers', required=False, nargs='*', help="Specific workers to get log files from")
+parser.add_argument('--take', required=False, type=int, default=None, help="Only download logs from the first N workers")
+
+
+args = parser.parse_args()
+
+is_local = not args.dest.startswith("gs://")
+
+if args.overwrite:
+    if is_local:
+        call('rm -r {dir}'.format(dir=args.dest), shell=True)
+    else:
+        call('gsutil -m rm -r {dir}'.format(dir=args.dest), shell=True)
+    
+
+master_dest = args.dest.rstrip('/') + "/master/"
+worker_dest = args.dest.rstrip('/') + "/workers/"
+
+if is_local:
+    call('mkdir -p {dir}'.format(dir=master_dest), shell=True)
+    call('mkdir -p {dir}'.format(dir=worker_dest), shell=True)
+
+desc = json.loads(Popen('gcloud dataproc clusters describe {name} --format json'.format(name=args.name),
+                        shell=True,
+                        stdout=PIPE,
+                        stderr=PIPE).communicate()[0].strip())
+
+config = desc['config']
+
+master = config['masterConfig']['instanceNames'][0]
+workers = config['workerConfig']['instanceNames']
+zone = re.search('zones\/(?P<zone>\S+)$', config['gceClusterConfig']['zoneUri']).group('zone')
+
+if args.workers:
+    invalid_workers = set(args.workers).difference(set(workers))
+    assert len(invalid_workers) == 0, "Non-existent workers specified: " + ", ".join(invalid_workers)
+    workers = args.workers
+
+if args.take:
+    assert args.take > 0 and args.take <= len(workers), "Number of workers to take must be in the range of [0, nWorkers]. Found " + args.take + "."
+    workers = workers[:args.take]
+
+    
+def gcloud_ssh(remote, command):
+    return 'gcloud compute ssh {remote} --zone {zone} --command "{command}"'.format(remote=remote, zone=zone, command=command)
+
+
+def gcloud_copy_files(remote, src, dest):
```

Comment:
Some of these functions look potentially useful for other purposes. Maybe wrap your script in a `main(args)` and keep these as separate functions that can be imported (might want to move them to a `cluster_utils.py` at some point in the future.